File: src/main/java/leetcode/permutaution/StringPermutations.java
package leetcode.permutaution;

import java.util.ArrayList;
import java.util.List;

public class StringPermutations {
    public static void main(String args[]) {
        permutation("123");
    }

    public static void permutation(String input) {
        List<String> result = new ArrayList<>();
        permutation(result, "", input);
        System.out.println(result);
    }

    private static void permutation(List<String> result, String perm, String word) {
        if (word.isEmpty()) {
            result.add(perm);
        } else {
            for (int i = 0; i < word.length(); i++) {
                permutation(result, perm + word.charAt(i), word.substring(0, i) + word.substring(i + 1));
            }
        }
    }
}

File: src/main/java/leetcode/permutaution/StringPermutations.kt
package leetcode.permutaution


fun main() {
    permutation("123")
}

fun permutation(input: String) {
    val result = ArrayList<String>()
    permutation(result, "", input)
    println(result)
}

private fun permutation(result: MutableList<String>, perm: String, word: String) {
    if (word.isEmpty()) {
        result.add(perm)
    } else {
        for (i in word.indices) {
            permutation(result, perm + word[i], word.substring(0, i) + word.substring(i + 1))
        }
    }
}